test/internal/layers/lazy.go

```diff
@@ -0,0 +1,165 @@
+}
+
+// Close removes the downloaded image layers.
+//
+// Does not take a [testing.TB] so it can be used in TestMain or init.
+func (x *LazyImageLayers) Close(ctx context.Context) error {
+	if x.dir == "" {
+		return nil
+	}
+
+	if _, err := os.Stat(x.dir); err != nil && !os.IsNotExist(err) {
+		return fmt.Errorf("path %q is not valid: %w", x.dir, err)
+	}
+	if err := wclayer.DestroyLayer(ctx, x.dir); err != nil {
+		return fmt.Errorf("could not destroy layer directory %q: %w", x.dir, err)
+	}
+	return nil
+}
+
+// Layers returns the image layer paths, from lowest to highest, for a particular image.
+func (x *LazyImageLayers) Layers(ctx context.Context, tb testing.TB) []string {
+	// basically combo of containerd fetch and unpack (snapshotter + differ)
+	tb.Helper()
+	var err error
+	x.once.Do(func() {
+		err = x.extractLayers(ctx)
+	})
+	if err != nil {
+		x.Close(ctx)
+		tb.Fatal(err)
+	}
+	return x.layers
+}
+
+// don't use tb.Error/Log inside Once.Do stack, since we cannot call tb.Helper before executing f()
+// within Once.Do and that will therefore show the wrong stack/location
+func (x *LazyImageLayers) extractLayers(ctx context.Context) (err error) {
+	log.G(ctx).Infof("pulling and unpacking %s image %q", x.Platform, x.Image)
+
+	if x.TempPath == "" {
+		dir := os.TempDir()
+		x.dir, err = os.MkdirTemp(dir, util.CleanName(x.Image))
+		if err != nil {
+			return fmt.Errorf("failed to create temp directory: %w", err)
+		}
+	} else {
+		x.dir, err = filepath.Abs(x.TempPath)
+		if err != nil {
+			return fmt.Errorf("failed to make %q absolute path: %w", x.TempPath, err)
+		}
+	}
+
+	var extract func(context.Context, io.ReadCloser, string, []string) error
+	switch x.Platform {
+	case constants.PlatformLinux:
+		extract = linuxImage
+	case constants.PlatformWindows:
+		if err = winio.EnableProcessPrivileges([]string{winio.SeBackupPrivilege, winio.SeRestorePrivilege}); err != nil {
+			return err
+		}
+		extract = windowsImage
+	default:
+		return fmt.Errorf("unsupported platform %q", x.Platform)
+	}
+
+	img, err := crane.Pull(x.Image, crane.WithPlatform(&v1.Platform{OS: x.Platform, Architecture: runtime.GOARCH}))
+	if err != nil {
+		return fmt.Errorf("failed to pull image %q: %w", x.Image, err)
+	}
+
+	layers, err := img.Layers()
+	if err != nil {
+		return fmt.Errorf("failed to get image %q layers: %w", x.Image, err)
+	}
+
+	for i, l := range layers {
+		d := filepath.Join(x.dir, strconv.FormatInt(int64(i), 10))
+		if err := os.Mkdir(d, 0755); err != nil {
+			return err
+		}
+		rc, err := l.Uncompressed()
+		if err != nil {
+			return fmt.Errorf("failed to load uncompressed layer for image %s: %w", x.Image, err)
+		}
+		defer rc.Close()
+		if err := extract(ctx, rc, d, x.layers); err != nil {
+			return fmt.Errorf("failed to extract layer %d for image %s: %w", i, x.Image, err)
+		}
+		x.layers = append(x.layers, d)
+	}
+
+	return nil
+}
+
+func linuxImage(ctx context.Context, rc io.ReadCloser, dir string, _ []string) error {
```

Comment:
Could we maybe refactor some of the code in dmverity-vhd to use in here too?

Reply:
as in try to combine them to share image pulling code?
I think that would require a lot of re-work, since `dmverity` allows specifying login info when pulling, and this code just relies on `crane` to do the right thing. Plus the code would need to be reworked to allow both linux and windows, like here, and the additional hashing needed for dmverity.

It should be possible for us to have our own internal code to pull and unpack for most situations, but i think that would be outside the scope of this PR

Reply:
Fair enough.